{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b4a1ae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "adata= sc.read_10x_mtx(\"/Users/shinjinichakraborty/desktop/filtered_feature_bc_matrix\")\n",
    "adata #check adata at this stage\n",
    "adata.obs #gives list of cells. It is a pandas dataframe\n",
    "adata.var #gives list of genes. It is a pandas dataframe\n",
    "adata.X   #sparse matrix combining cell and gene data\n",
    "\n",
    "#adding columns to the object\n",
    "adata.obs[\"cell_label\"]=\"cell_xx\"\n",
    "adata.obs #checking if new column is added to the object\n",
    "\n",
    "#next step is quality control.This will include filtering out low reads in cells and similar noise! \n",
    "#This is different from what cell-ranger does.\n",
    "#low gene expression, less cells with particular gene expressed and mitochondrial gene\n",
    "\n",
    "#cells with low gene expression - preprocessing -pp\n",
    "\n",
    "sc.pp.filter_cells(adata,min_genes=50)\n",
    "adata #check if adata has filtered genes excluded\n",
    "\n",
    "#filter low gene expression - preprocessing -pp\n",
    "\n",
    "sc.pp.filter_genes(adata,min_cells=3)\n",
    "adata #check if adata has filtered cells excluded\n",
    "\n",
    "#filter mitochondrial content\n",
    "adata.obs['percent_mito'] = np.sum(adata[:, adata.var['mt'].values].X, axis=1).A1 / np.sum(adata.X, axis=1).A1\n",
    "adata.obs #to check if 'percent mito' is added to the object dataframe\n",
    "sc.pl.violin(adata, 'percent_mito', jitter=0.4) #gives violin plot, giving an idea what threshold can be suitably set\n",
    "adata = adata[adata.obs['percent_mito'] < 0.3, :] #this is called boolean filtering\n",
    "adata #check if adata has filtered mitochondrial genes\n",
    "\n",
    "# Normalization has two steps - preprocessing normalization and log transformation\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1e4) #normalization\n",
    "sc.pp.log1p(adata) #log transformation\n",
    "adata #Anndata object number remains the same, but an additional layer is added when log transformed\n",
    "\n",
    "# Downstream analysis starts with dimensionality reduction. First step is PCA\n",
    "# And then, PCA output used as the input for tSNE and UMAP\n",
    "\n",
    "sc.tl.pca(adata, svd_solver='arpack')\n",
    "\n",
    "adata #several other layers added\n",
    "\n",
    "sc.pl.pca(adata, color='CST3')\n",
    "\n",
    "#Now we perform UMAP\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40) # this computes neighborhood graph required for UMAP\n",
    "sc.tl.umap(adata)\n",
    "# Run Louvain clustering\n",
    "sc.tl.louvain(adata)\n",
    "# Plot PCA colored by Louvain clusters\n",
    "sc.pl.pca(adata, color='louvain')\n",
    "\n",
    "# We can plot gene specific UMAPs as well \n",
    "\n",
    "sc.pl.pca_variance_ratio(adata, log=True)\n",
    "sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n",
    "sc.tl.paga(adata)\n",
    "sc.pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\n",
    "sc.tl.umap(adata, init_pos='paga')\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=['CST3', 'NKG7', 'PPBP'])\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
